app/workers/metrics/record_db_and_es_record_counts_worker.rb

```diff
@@ -0,0 +1,64 @@
+# This worker checks the number of records for each Elasticsearch index and
+# compares it to the number of records in our database.
+#
+# For indexes that don't match a single model (no model or multiple models) we
+# need to implement a custom db_count method on the Search class to do the
+# counts for us
+module Metrics
+  class RecordDbAndEsRecordCountsWorker
+    include Sidekiq::Worker
+
+    sidekiq_options queue: :low_priority, retry: 5
+
+    # Once the Elasticsearch migration is complete we can update this value to
+    # Search::Cluster::SEARCH_CLASSES which will include all indexes.
+    SEARCH_CLASSES = [
+      Search::ChatChannelMembership,
+      Search::ClassifiedListing,
+      Search::Tag,
+    ].freeze
+
+    # Adjustable margin of error - this is how far off the index count can be
+    # from the database count before we raise an error
+    def perform(margin_of_error = 0)
+      SEARCH_CLASSES.each do |search_class|
+        margin_of_error = margin_of_error.to_i
+        db_count = db_count(search_class)
```

Comment:
I know you haven't implemented this yet, but `COUNT()` in Postgres always does a full-table scan, which we may not want to do for several big models in a loop. @rhymes added an `estimated_count` method to `ApplicationRecord`, maybe we should try to use that.

* https://www.citusdata.com/blog/2016/10/12/count-performance/#dup_counts_estimated_full
* https://stackoverflow.com/a/48391562/4186181

Reply:
Ooooo this is very interesting - TIL 🤔.

I don't think it's going to work since we need the exact count and we're racing against the clock. I tried `estimated_count` locally too and the counts were off. I could see that difference leading to false failures.

Reply:
I think this ties to my comment above re using a percentage instead of an absolute value as an indicator for drift.

Reply:
@atsmith813 `.estimated_count` locally is off because we have a "toy DB" and the vacuum cleaner doesn't really work. Same for the stats collector. It's the same reason why it's hard to make accurate estimates about query plans in `development` as our DB has order of magnitudes less data in some tables than production.

I worry `db_count` might end up slowing down the site eventually. It might not be a problem for now as I don't think any of the searchable models are tied to massive tables. According to the API - `curl https://dev.to/api/articles | jq '.[0].id'` we have 286939 published articles and that's probably the biggest table.

I just tried to run the query `ApplicationRecord.estimated_count` runs using Blazer:

```sql
SELECT (
  (reltuples / GREATEST(relpages, 1)) *
  (pg_relation_size('articles') / (GREATEST(current_setting('block_size')::integer, 1)))
)::bigint AS count
FROM pg_class WHERE relname = 'articles';
```

and it returns 0 but I think it's because Blazer doesn't allow me to query that table. 
I can see it works in production since https://dev.to/partnerships displays the estimated count of users.

You might want to verify using the production console to check how off the count is for articles.

If you're confident about what `.estimated_count` please use that plus a percentage as suggested by @citizen428 which should lead to less errors in monitoring

Reply:
Ahhhh, I see. That makes sense, thank you! TIL.

I don't have access to a production console so I have no idea. I'm hesitant about percentages 🤔 I'm thinking I might save that for the future.

My thinking right now is to get this monitor in place to help with verifying the new search integrations to help catch bugs like this one https://github.com/thepracticaldev/dev.to/pull/6517 sooner. We caught that because the count was off by 1. Granted, over time, that count would've grown and eventually would've passed some threshold. But I'm thinking of keeping the margin tight while we're implementing new searches and then widening it from there. The `FeedContent` search that Molly is currently working on is the biggest and will be a good test of noise. I'm thinking once we take Algolia out we'll have a sense of sensitivity for this one and can move to percentages.

To your point about Articles likely being the biggest, it's possible we may want to adjust the sensitivity by index too, which would be implemented differently.

We also already have a gap to account for with timing. If we add an estimated count then the gap will become even wider and less accurate.

This task runs once a day with 5 retries. Maybe we schedule it to run in off-peak hours? So that translates to 5 queries max, per index, per day. Would that help it at all?